### PR TITLE
cpu/cc2538: add driver for the MCU's integrated radio

### PR DESCRIPTION
**Pull request by hexluthor**

### CODE REVIEW

File: cpu/cc2538/include/cc2538-rf.h [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+#define CC2538_RF_FIFO_SIZE         128
+#define CC2538_FCS_LEN                2
+#define CC2538_AUTOCRC_LEN            2
+
+#define CC2538_RF_MAX_PACKET_LEN    (CC2538_RF_FIFO_SIZE - 1)
+
+#define CC2538_MAX_DATA_LENGTH      CC2538_RF_MAX_PACKET_LEN
+
+#define rfcore_assert(expr) (void)( (expr) || rfcore_assert_failure(#expr, __FUNCTION__, __LINE__) )
```
Comment (thomaseichinger) [inline cpu/cc2538/include/cc2538-rf.h]: This and all following, please capitalise the macro name.

File: cpu/cc2538/radio/cc2538-rf-netdev.c [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    }
+}
+
+static int get_state(netdev_t *dev, netdev_state_t *state)
+{
+    if (dev != &cc2538_netdev) {
+        return -ENODEV;
+    }
+
+    return netdev_state;
```
Comment (thomaseichinger) [inline cpu/cc2538/radio/cc2538-rf-netdev.c]: `netdev_state` should be returned through `state`.

File: cpu/cc2538/radio/cc2538-rf-netdev.c [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+        case NETDEV_STATE_POWER_IDLE:
+            cc2538_off();
+            break;
+
+        case NETDEV_STATE_RX_MODE:
+            cc2538_switch_to_rx();
+            break;
+
+        case NETDEV_STATE_PROMISCUOUS_MODE:
+            cc2538_set_monitor(false);
```
Comment (thomaseichinger) [inline cpu/cc2538/radio/cc2538-rf-netdev.c]: I think this should be `cc2538_set_monitor(true);`.

File: cpu/cc2538/radio/cc2538-rf-netdev.c [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+            hdr.short_addrs.dest_pan  = HTONS(dest->pan.id);
+            hdr.short_addrs.dest_addr = HTONS(dest->pan.addr);
+            hdr.short_addrs.src_pan   = HTONS(src_pan);
+            hdr.short_addrs.src_addr  = HTONS(cc2538_get_address());
+            hdr_size = 3 + sizeof(short_addrs_t);
+        }
+    }
+
+    /* total frame size */
+    if (1 + hdr_size + len > CC2538_RF_FIFO_SIZE) {
```
Comment (thomaseichinger) [inline cpu/cc2538/radio/cc2538-rf-netdev.c]: Please document `1` or provide a `#define`. I'd guess it is for `CC2538_AUTOCRC_LEN`?

**Comment by hexluthor** [comment]
Thanks for the feedback @thomaseichinger. I believe I have addressed your concerns.

File: cpu/cc2538/radio/cc2538-rf.c [hunk 5]
```diff
@@ -0,0 +1,854 @@
...
+            hdr.short_addrs.dest_pan  = HTONS(dest.pan.id);
+            hdr.short_addrs.dest_addr = HTONS(dest.pan.addr);
+            hdr.short_addrs.src_pan   = HTONS(src_pan);
+            hdr.short_addrs.src_addr  = HTONS(cc2538_get_address());
+            hdr_size = 3 + sizeof(short_addrs_t);
+        }
+    }
+
+    /* total frame size */
+    if (CC2538_RF_FIFO_SIZE + hdr_size + len > CC2538_RF_FIFO_SIZE) {
```
Comment (thomaseichinger) [inline cpu/cc2538/radio/cc2538-rf.c]: `s/CC2538_RF_FIFO_SIZE/CC2538_PACKET_LENGTH_SIZE/`?

**Comment by thomaseichinger** [comment]
@hexluthor please consider https://github.com/locicontrols/RIOT/pull/1 to pass build tests.

**Comment by thomaseichinger** [comment]
@hexluthor Sorry for the delay. Could you please rebase? I finally got the OpenMote board to test this.

**Comment by hexluthor** [comment]
Do I need to squash this, or is that done during the merge?

**Comment by thomaseichinger** [comment]
- Travis fails because the b4181f5 commit message triggers a script checking commit messages.

- You will need to squash before we merge it.

- RX:
There is a problem using the default example as `cc2538_packet_t` differs from the other radio packet types. Output: 
```bash
2015-01-21 10:40:03,324 - INFO # Got radio packet:
2015-01-21 10:40:03,324 - INFO #        Length: 18
2015-01-21 10:40:03,325 - INFO #        Src:    0
2015-01-21 10:40:03,325 - INFO #        Dst:    0
2015-01-21 10:40:03,326 - INFO #        LQI:    107
2015-01-21 10:40:03,326 - INFO #        RSSI:   24
2015-01-21 10:40:03,327 - INFO # Payload Length:0
2015-01-21 10:40:03,327 - INFO # Payload:A�
```
- TX:
Transmission seems to work except the address fields seem to be incorrect. Receiving (on a different board):
```bash
2015-01-21 11:04:46,067 - INFO # > Got radio packet:
2015-01-21 11:04:46,067 - INFO #        Length: 18
2015-01-21 11:04:46,068 - INFO #        Src:    0
2015-01-21 11:04:46,068 - INFO #        Dst:    0
2015-01-21 11:04:46,068 - INFO #        LQI:    69
2015-01-21 11:04:46,069 - INFO #        RSSI:   3
2015-01-21 11:04:46,070 - INFO # Payload Length:13
2015-01-21 11:04:46,070 - INFO # Payload:jsdlkajsldad
```

**Comment by thomaseichinger** [comment]
@hexluthor You basically implemented 3 different radio driver interfaces.

1. `radio_driver.h` This one was introduced but never really used except for the cc2420.
2. `transceiver.h` Which is currently used but will be replaced by `netdev`.
3. netdev interface (based on concepts from `radio_driver.h`) which is part of our ongoing work to refactor the network stack.

I'm not objecting against implementing `radio_driver.h` but it will be removed soon. I'm sorry for the confusion.

**Comment by hexluthor** [comment]
@thomaseichinger I think I may have fixed the default-example issues, but more testing is needed. No worries about the three different interfaces. I wasn't sure which would be more future-proof so I took a shotgun approach. But thank you for clarifying!

**Comment by thomaseichinger** [comment]
@hexluthor 
- TX doesn't work at all now. I think [transceiver.c:L844](https://github.com/locicontrols/RIOT/blob/cc2538-radio/sys/transceiver/transceiver.c#L844) needs to be changed.

- RX seems quite good now. Three remaining issues,
  1. The RSSI value seems to overflow. Does cc2538 provide a signed value for this?
  2. The last payload byte sent gets lost. Sent payloads are `lkjasljkdaklds`, `hello`, `hellocc2538`.
  3. One payload buffer seems to contain old values.

```bash
2015-01-22 10:30:41,413 - INFO # Got radio packet:
2015-01-22 10:30:41,413 - INFO #        Length: 24
2015-01-22 10:30:41,414 - INFO #        Src:    1
2015-01-22 10:30:41,414 - INFO #        Dst:    2
2015-01-22 10:30:41,415 - INFO #        LQI:    106
2015-01-22 10:30:41,415 - INFO #        RSSI:   4294967281
2015-01-22 10:30:41,416 - INFO # Payload Length:13
2015-01-22 10:30:41,416 - INFO # Payload:lkjasljkdakld

2015-01-22 10:53:00,791 - INFO # Got radio packet:
2015-01-22 10:53:00,812 - INFO #        Length: 15
2015-01-22 10:53:00,812 - INFO #        Src:    1
2015-01-22 10:53:00,813 - INFO #        Dst:    2
2015-01-22 10:53:00,813 - INFO #        LQI:    106
2015-01-22 10:53:00,814 - INFO #        RSSI:   24
2015-01-22 10:53:00,814 - INFO # Payload Length:4
2015-01-22 10:53:00,815 - INFO # Payload:hellsljkdakld

2015-01-22 10:53:13,265 - INFO # Got radio packet:
2015-01-22 10:53:13,266 - INFO #        Length: 21
2015-01-22 10:53:13,266 - INFO #        Src:    1
2015-01-22 10:53:13,286 - INFO #        Dst:    2
2015-01-22 10:53:13,287 - INFO #        LQI:    106
2015-01-22 10:53:13,287 - INFO #        RSSI:   24
2015-01-22 10:53:13,288 - INFO # Payload Length:10
2015-01-22 10:53:13,289 - INFO # Payload:hellocc253kld
```

**Comment by hexluthor** [comment]
All fixed, I think. The cc2538 doesn't provide a signed RSSI, but it is offset by 73 dB. I subtract-out the offset now and modified `sc_transceiver.c` to print a signed value (`ieee802154_packet_t::rssi` is signed). The default example tests ok on my cc2538dk's.

**Comment by hexluthor** [comment]
Travis seems broken. Should I go ahead and squash the commits?

**Comment by cgundogan** [comment]
This PR needs a rebase first

**Comment by hexluthor** [comment]
Rebased.

**Comment by authmillenon** [comment]
The `netdev` version you implemented against is deprecated. Please refer to [`ng_netdev.h`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/ng_netdev.h) and the [Wiki](https://github.com/RIOT-OS/RIOT/wiki/How-to-implement-a-%60ng_netdev%60-enabled-device).

**Comment by OlegHahm** [comment]
@hexluthor (or anyone else) cares to update against current `netdev` version?

**Comment by hexluthor** [comment]
My eyes are spinning. Should this be implemented as a `gnrc_netdev_driver_t` or a `netdev2_driver_t`?

**Comment by OlegHahm** [comment]
Please go for netdev2. This is supposed to be the default interface in the long run.

**Comment by hexluthor** [comment]
@AlexanderS, @aeneby, & others: here is my WIP for a `netdev2` driver for the cc2538. I don't have time to work on it right now, but I wanted to get the code out there so others can work with it.

**Comment by OlegHahm** [comment]
Thanks for sharing this, @hexluthor. Maybe @A-Paul is interested, too.

**Comment by A-Paul** [comment]
@OlegHahm,
> Thanks for sharing this, @hexluthor. Maybe @A-Paul is interested, too.

I'm an unlikely candidate to work on the code.

**Comment by A-Paul** [comment]
Set WIP flag, according to @hexluthor's comment.

**Comment by aeneby** [comment]
@A-Paul @hexluthor I'm actually working on a new implementation of a cc2538 RF driver, based loosely on this PR (didn't really make sense to continue with this one since so much has changed in RIOT network-wise since it was written). I'm just waiting on a couple of PRs (like #4645 and #4636), at which point I'll complete basic functionality and create a new PR.

**Comment by A-Paul** [comment]
Hi @aeneby, thank you for giving feedback. I'm looking forward to your implementation.
When you create the new PR, please reference this one and leave a comment here.

**Comment by aeneby** [comment]
A more complete driver implementation here: #5291

**Comment by OlegHahm** [comment]
Can we close this one then?

**Comment by hexluthor** [comment]
Closing in favor of https://github.com/RIOT-OS/RIOT/pull/5291 . Thanks for your efforts @aeneby!

**Pull request closed without merging**